flask/launcher/src/main/java/net/corda/flask/launcher/JarCache.java

```diff
@@ -0,0 +1,226 @@
+                    ioe
+            );
+            return false;
+        }
+    }
+
+    @SneakyThrows
+    private static Path computeCacheDirectory(String appName) {
+        Stream<Optional<Path>> candidates;
+        if (OS.isUnix) {
+            candidates = Stream.of(
+                    Optional.ofNullable(System.getProperty(Flask.JvmProperties.CACHE_DIR)).map(Paths::get),
+                    Optional.ofNullable(System.getenv("XDG_CACHE_HOME")).map(prefix -> Paths.get(prefix, appName)),
+                    Optional.ofNullable(System.getProperty("user.home")).map(prefix -> Paths.get(prefix, ".cache", appName)),
+                    Optional.ofNullable(System.getProperty("java.io.tmpdir")).map(path -> Paths.get(path).resolve(appName)),
+                    Optional.of(Paths.get("/tmp", appName))
+            );
+        } else if (OS.isMac) {
+            candidates = Stream.of(
+                    Optional.ofNullable(System.getProperty(Flask.JvmProperties.CACHE_DIR)).map(Paths::get),
+                    Optional.ofNullable(System.getProperty("user.home"))
+                            .map(prefix -> Paths.get(prefix, "Library", "Saved Application State", appName)),
+                    Optional.of(Paths.get("/Library/Caches", appName)),
+                    Optional.ofNullable(System.getProperty("java.io.tmpdir"))
+                            .map(prefix -> Paths.get(prefix).resolve(appName)));
+        } else if (OS.isWindows) {
+            candidates = Stream.of(
+                    Optional.ofNullable(System.getProperty(Flask.JvmProperties.CACHE_DIR)).map(Paths::get),
+                    Optional.ofNullable(System.getenv())
+                            .map(it -> it.get("LOCALAPPDATA"))
+                            .map(prefix -> Paths.get(prefix, appName)),
+                    Optional.ofNullable(System.getProperty("user.home"))
+                            .map(prefix -> Paths.get(prefix, "AppData", "Local", appName)),
+                    Optional.ofNullable(System.getProperty("user.home"))
+                            .map(prefix -> Paths.get(prefix, "Local Settings", "Application Data", appName)),
+                    Optional.ofNullable(System.getProperty("java.io.tmpdir"))
+                            .map(prefix -> Paths.get(prefix).resolve(appName))
+            );
+
+        } else {
+            candidates = Stream.of(
+                    Optional.ofNullable(System.getProperty(Flask.JvmProperties.CACHE_DIR)).map(Paths::get),
+                    Optional.ofNullable(System.getProperty("java.io.tmpdir")).map(path -> Paths.get(path).resolve(appName))
+            );
+        }
+        return candidates
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .filter(JarCache::validateCacheDirectory)
+                .findFirst()
+                .orElseThrow(() -> new FileNotFoundException("Unable to find a usable cache directory"));
```

Comment:
Or you could just _create_ the cache directory in `${user.home}/.flask` on _all_ operating systems. If `user.home` is `null` then I suspect that means you're actually running inside a Docker container and so falling back to a subdirectory of `${java.io.tmpdir}` would be fine.

The point about defining and creating your own directory is that then you _know_ that it should have the properties that you need - i.e. "is writable". And if it _isn't_ writable for any reason (e.g. the OS detects an errors and makes the entire file-system "read-only") then you know that something is wrong and can exit.

Reply:
yeah, I'll do that